File: runtime/python/units/wasi/fd_close.py
def wasi_fd_close(self, fd):
    io = self.fds.pop(fd, None)
    if io is None:
        return self.ERRNO_BADF
    if isinstance(io, self.WasiDir):
        return self.ERRNO_SUCCESS  # no real OS handle to close
    if io not in self.std_ios:
        io.close()
    return self.ERRNO_SUCCESS

File: runtime/perl/units/wasi/fd_close.pl
# requires:
sub wasi_fd_close {
    my ($self, $fd) = @_;
    my $e = delete $self->{fds}{$fd};
    return ERRNO_BADF unless defined $e;
    # A dir has no real OS handle to close; stdio is never closed.
    close($e->{fh}) if !$e->{dir} && !defined($e->{std});
    return ERRNO_SUCCESS;
}
